lib/pharos/kubeadm/cluster_config.rb

```diff
@@ -0,0 +1,208 @@
+
+      # @param config [Pharos::Config] cluster config
+      # @param host [Pharos::Configuration::Host] master host-specific config
+      def initialize(config, host)
+        @config = config
+        @host = host
+      end
+
+      # @return [Hash]
+      def generate
+        config = {
+          'apiVersion' => 'kubeadm.k8s.io/v1alpha3',
+          'kind' => 'ClusterConfiguration',
+          'kubernetesVersion' => Pharos::KUBE_VERSION,
+          'imageRepository' => @config.image_repository,
+          'apiServerCertSANs' => build_extra_sans,
+          'networking' => {
+            'serviceSubnet' => @config.network.service_cidr,
+            'podSubnet' => @config.network.pod_network_cidr
+          },
+          'controlPlaneEndpoint' => 'localhost:6443', # client-side loadbalanced kubelets
+          'apiServerExtraArgs' => {},
+          'controllerManagerExtraArgs' => {
+            'horizontal-pod-autoscaler-use-rest-clients' => 'true'
+          }
+        }
+
+        if @config.cloud && @config.cloud.provider != 'external'
+          config['apiServerExtraArgs']['cloud-provider'] = @config.cloud.provider
+          config['controllerManagerExtraArgs']['cloud-provider'] = @config.cloud.provider
+          if @config.cloud.config
+            config['apiServerExtraArgs']['cloud-config'] = CLOUD_CFG_FILE
+            config['controllerManagerExtraArgs']['cloud-config'] = CLOUD_CFG_FILE
+          end
+        end
+
+        # Only configure etcd if the external endpoints are given
+        if @config.etcd&.endpoints
+          configure_external_etcd(config)
+        else
+          configure_internal_etcd(config)
+        end
+        config['apiServerExtraVolumes'] = [
+          {
+            'name' => 'pharos',
+            'hostPath' => PHAROS_DIR,
+            'mountPath' => PHAROS_DIR
+          }
+        ]
+
+        config['controllerManagerExtraVolumes'] = [
+          {
+            'name' => 'pharos',
+            'hostPath' => PHAROS_DIR,
+            'mountPath' => PHAROS_DIR
+          }
+        ]
+
+        # Only if authentication token webhook option are given
+        configure_token_webhook(config) if @config.authentication&.token_webhook
+
+        # Configure audit related things if needed
+        configure_audit_webhook(config) if @config.audit&.webhook&.server
+        configure_audit_file(config) if @config.audit&.file
+
+        configure_admission_plugins(config)
+
+        # Set secrets config location and mount it to api server
+        config['apiServerExtraArgs']['experimental-encryption-provider-config'] = SECRETS_CFG_FILE
+        config['apiServerExtraVolumes'] << {
+          'name' => 'k8s-secrets-config',
+          'hostPath' => SECRETS_CFG_DIR,
+          'mountPath' => SECRETS_CFG_DIR
+        }
+
+        config
+      end
+
+      # @return [Array<String>]
+      def build_extra_sans
+        extra_sans = Set.new(['localhost'])
```

Comment:
Rather unnecessary use of Set here and also `require 'set'` is needed when not running via bundler.

Reply:
I think only reason for `Set` is uniqueness of addresses.